### PR TITLE
EOS-16391: Increase parallelization in accessing cob domain btrees

### PR DESCRIPTION
**Pull request by tejalsheth**
Added an array (currently size 16) for btree cd_object_index, cd_namespace, cd_fileattr_basic,
and cd_fileattr_ea in structure m0_cob_domain.

Used m0_fid_hash() to get the hash on the unique fid from the key of each struct
mentioned above. Then did mod operation (currently %16) on the returned hash
to use that as array index.

Signed-off-by: Tejal Sheth <tejal.sheth@seagate.com>

### CODE REVIEW

File: cob/cob.c [hunk 1]
```diff
@@ -478,27 +478,33 @@ M0_UNUSED static char *cob_dom_id_make(char *buf, const struct m0_cob_domain_id
...
 	M0_PRE(dom != NULL);
 	M0_PRE(dom->cd_id.id != 0);
-
-	m0_be_btree_init(&dom->cd_object_index,   seg, &cob_oi_ops);
-	m0_be_btree_init(&dom->cd_namespace,	  seg, &cob_ns_ops);
-	m0_be_btree_init(&dom->cd_fileattr_basic, seg, &cob_fab_ops);
+	for(i = 0; i < COB_HT_SIZE; i++)
+	{ 
+		m0_be_btree_init(&dom->cd_object_index[i],   seg, &cob_oi_ops);
+		m0_be_btree_init(&dom->cd_namespace[i],	  seg, &cob_ns_ops);
```
Comment (huanghua78) [inline cob/cob.c:489]: Alignment.

File: cob/cob.c [hunk 2]
```diff
@@ -1669,21 +1709,24 @@ M0_INTERNAL int m0_cob_name_del(struct m0_cob *cob,
...
+	ht_idx = cob_get_hash(&nskey->cnk_pfid);
 
 	/*
 	 * Kill name from namespace.
 	 */
 	m0_buf_init(&key, nskey, m0_cob_nskey_size(nskey));
 	m0_buf_init(&val, &nsrec, sizeof nsrec);
 
-	rc = cob_table_lookup(&cob->co_dom->cd_namespace, &key, &val);
+
```
Comment (huanghua78) [inline cob/cob.c:1724]: Remove extra blank line.

File: cob/cob.c [hunk 3]
```diff
@@ -1708,34 +1752,41 @@ M0_INTERNAL int m0_cob_name_update(struct m0_cob *cob,
 	struct m0_buf        val;
 	int                  rc;
 
+	uint16_t            ht_srcidx;
+	uint16_t            ht_tgtidx;
+	uint16_t            ht_idx;
```
Comment (huanghua78) [inline cob/cob.c:1757]: These variables are used only once.
1. Can we use a single variable?
2. Can we:
```
cob_table_lookup(&cob->co_dom->cd_namespace[cob_get_hash(&srckey->cnk_pfid)], &key, &val);
```

File: cob/cob.c [hunk 4]
```diff
@@ -1992,41 +2050,41 @@ M0_INTERNAL void m0_cob_tx_credit(struct m0_cob_domain *dom,
 		break;
 	case M0_COB_OP_CREATE:
 		m0_cob_tx_credit(dom, M0_COB_OP_NAME_ADD, accum);
-		TCREDIT(&dom->cd_fileattr_basic, INSERT, FAB, accum);
+		TCREDIT(&dom->cd_fileattr_basic[0], INSERT, FAB, accum);
```
Comment (huanghua78) [inline cob/cob.c:2053]: Please add comments to say that:
When calculating credit, we can just use the first element in the array.
The result will be the same.

File: cob/cob.h [hunk 5]
```diff
@@ -267,11 +269,11 @@ struct m0_cob_domain {
...
 	 */
-	struct m0_be_btree      cd_object_index;
-	struct m0_be_btree      cd_namespace;
-	struct m0_be_btree      cd_fileattr_basic;
+	struct m0_be_btree      cd_object_index[COB_HT_SIZE];
+	struct m0_be_btree      cd_namespace[COB_HT_SIZE];
+	struct m0_be_btree      cd_fileattr_basic[COB_HT_SIZE];
 	struct m0_be_btree      cd_fileattr_omg;
-	struct m0_be_btree      cd_fileattr_ea;
+	struct m0_be_btree      cd_fileattr_ea[COB_HT_SIZE];
```
Comment (huanghua78) [inline cob/cob.h:276]: Please add comments to explain why the cd_fileattr_omg does not need to be an array.
* OMG will be relatively a small set of data. No need to be divided into multiple btree?

File: be/tool/beck.c [hunk 6]
```diff
@@ -2979,9 +2979,9 @@ static void cob_act(struct action *act, struct m0_be_tx *tx)
 	struct cob_action        *ca = container_of(act, struct cob_action,
 						    coa_act);
 	struct m0_be_btree       *ios_ns = &act->a_builder->b_ios_cdom->
-								cd_namespace;
+								cd_namespace[0];
 	struct m0_be_btree       *mds_ns = &act->a_builder->b_mds_cdom->
-								cd_namespace;
+								cd_namespace[0];
```
Comment (huanghua78) [inline be/tool/beck.c:2984]: Please explain why the first element of the cd_namespace array is used.

File: cob/cob.h [hunk 7]
```diff
@@ -230,6 +230,8 @@ enum {
 	M0_IOS_COB_ID_START = 1000,
 };
 
+enum { COB_HT_SIZE = 16 };
```
Comment (sg-shankar) [inline cob/cob.h:227]: The size of hash should be same as emap, 256?

**Comment by kanchan-chaudhari** [comment]
Closing this PR as this would be merged with main after validating stability and performance along with cas parallelization changes.

**Pull request closed without merging**